### PR TITLE
refactor: prepare html-reporter for pwt GUI integration

### PR DESCRIPTION
**Pull request by shadowusr**
### What's done?
- re-written hermione.js to typescript
- renamed sqlite-adapter to a more precise sqlite-client
- got rid of init method in sqlite-client
- began to use ReporterTestResult instead of tool-specific/weird types in sqlite-adapter
- made attempt field of test result readonly and introduced explicit handling of attempt number

### Why?
- These changes make the code more explicit, simple and clear, in my opinion
- These changes open a road to pwt GUI integration

⚠️ I recommend reviewing this PR commit-by-commit.

### CODE REVIEW

File: hermione.ts [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+    });
+};
+
+async function prepare(hermione: Hermione, reportBuilder: StaticReportBuilder, pluginConfig: ReporterConfig): Promise<void> {
+    const {path: reportPath} = pluginConfig;
+    const {imageHandler} = reportBuilder;
+
+    const failHandler = async (testResult: HermioneTestResult): Promise<ReporterTestResult> => {
+        const attempt = reportBuilder.testAttemptManager.registerAttempt({fullName: testResult.fullTitle(), browserId: testResult.browserId}, FAIL);
+        const formattedResult = formatTestResult(testResult, FAIL, attempt, reportBuilder);
```
[the remark this answers is not in the document]
Comment (shadowusr) [reply hermione.ts:42]: Я бы здесь оставил аргументы. На мой взгляд, до 4-5 аргументов — это норм. Если больше — стоит объединять. Но очень не хочется иметь лишние сущности и типы в виде testOpts. Я как раз стараюсь везде от этого уходить — десятков маленьких объектов, у которых разные свойства. Поэтому, если ты не настаиваешь, я бы пока что оставил.

File: hermione.ts [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+            actions.push((formattedResult as HermioneTestAdapter).saveErrorDetails(reportPath));
+        }
+
+        await Promise.all(actions);
+
+        return formattedResult;
+    };
+
+    const addFail = (formattedResult: ReporterTestResult): ReporterTestResult => {
+        return hasDiff(formattedResult.assertViewResults as {name?: string}[])
```
[the remark this answers is not in the document]
Comment (shadowusr) [reply hermione.ts:56]: ну да, это не совсем удобно, но зашито в sqlite базки. поэтому тут если менять что-то будем, то либо накручивать адаптеры, либо развалится старый формат базок.

File: lib/plugin-adapter.ts [hunk 3]
```diff
@@ -56,18 +58,23 @@ export class PluginAdapter {
     }
 
     protected async _createStaticReportBuilder(prepareData: PrepareFn): Promise<void> {
-        const staticReportBuilder = StaticReportBuilder.create(this._hermione.htmlReporter, this._config);
+        if (!this._hermione.htmlReporter) {
+            throw new Error('Html-reporter API has to be added to hermione before usage');
```
[the remark this answers is not in the document]
Comment (shadowusr) [reply lib/plugin-adapter.ts:62]: Выгасил `PluginAdapter`, и проблема ушла.

File: lib/plugin-adapter.ts [hunk 4]
```diff
@@ -56,18 +58,23 @@ export class PluginAdapter {
...
+        }
 
-        await staticReportBuilder.init();
+        const dbClient = await SqliteClient.create({htmlReporter: this._hermione.htmlReporter, reportPath: this._config.path});
+        const testAttemptManager = new TestAttemptManager();
```
[the remark this answers is not in the document]
Comment (shadowusr) [reply lib/plugin-adapter.ts:66]: Потому что testAttemptManager может существовать без reportBuilder и использоваться в других местах. Ну и чтобы следовать dependency injection со всеми вытекающими плюсами (простота тестирования, гибкость замены на другую реализацию)

File: lib/gui/tool-runner/index.ts [hunk 5]
```diff
@@ -298,9 +308,13 @@ export class ToolRunner {
             const testId = formatId(test.id.toString(), browserId);
             this._tests[testId] = _.extend(test, {browserId});
 
-            test.pending
-                ? reportBuilder.addSkipped(formatTestResultUnsafe(test, SKIPPED, reportBuilder))
-                : reportBuilder.addIdle(formatTestResultUnsafe(test, IDLE, reportBuilder));
+            if (test.pending) {
+                const attempt = reportBuilder.testAttemptManager.registerAttempt({fullName: test.fullTitle(), browserId: test.browserId}, SKIPPED);
+                reportBuilder.addSkipped(formatTestResultUnsafe(test, SKIPPED, attempt, reportBuilder));
```
[the remark this answers is not in the document]
Comment (shadowusr) [reply lib/gui/tool-runner/index.ts:313]: При текущем подходе не можем по следующим причинам:
- везде в приложении, включая `addSkipped` мы условились работать с `ReporterTestResult`
- для создания `ReporterTestResult` нужен `attempt`

Можно было бы поместить его в фабрику `formatTestResult`, но тогда каждый раз, когда хотим отформатировать testResult считался бы ретраем, а это — неожиданный сайд-эффект.

В целом я задумывал `testAttemptManager` как независимый сервис по отслеживанию попыток теста, можно думать про него, что это база данных. Сам результат теста не знает, какой он по счету, в каждом инстансе `ReporterTestResult` зафиксирована конкретная попытка и он тоже не знает о глобальном состоянии. А `testAttemptManager` — как раз глобальное хранилище состояния попыток.

Если есть идеи, как можно упростить/улучшить этот момент — буду рад услышать!

File: hermione.ts [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+            : reportBuilder.addError(formattedResult);
+    };
+
+    return new Promise((resolve, reject) => {
+        const queue = new PQueue({concurrency: os.cpus().length});
+        const promises: Promise<unknown>[] = [];
+
+        hermione.on(hermione.events.TEST_PASS, testResult => {
+            promises.push(queue.add(async () => {
+                const attempt = reportBuilder.testAttemptManager.registerAttempt({fullName: testResult.fullTitle(), browserId: testResult.browserId}, FAIL);
```
[the remark this answers is not in the document]
Comment (shadowusr) [reply hermione.ts:67]: Мне кажется, что прокси-методы, которые просто торчат наружу — не очень практика:
- есть дублирование со всеми вытекающими
- пытаемся сымитировать наследование, что противоречит [composition over inheritance](https://en.wikipedia.org/wiki/Composition_over_inheritance)

На самом деле он и торчать из reportBuilder'а не должен, просто так временно удобно его было передавать по функциям. По хорошему я вижу, что это должно быть приватное свойство, а где нужно он должен передаваться отдельно

Что думаешь?

File: lib/test-attempt-manager.ts [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+
+        if (_.isNil(attempt)) {
+            this._attemptMap.set(hash, 0);
+            statuses.push(status);
+            this._statusMap.set(hash, statuses);
+
+            return 0;
+        }
+
+        if ([IDLE, RUNNING, SKIPPED, UPDATED].includes(_.last(statuses) as TestStatus)) {
```
[the remark this answers is not in the document]
Comment (shadowusr) [reply lib/test-attempt-manager.ts:56]: Да, это проблема. Спасибо за edge-case!

File: lib/test-attempt-manager.ts [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+
+const INVALID_ATTEMPT = -1;
+
+export class TestAttemptManager {
+    private _attemptMap: Map<string, number>;
+    private _statusMap: Map<string, TestStatus[]>;
+
+    constructor() {
+        this._attemptMap = new Map();
+        this._statusMap = new Map();
```
[the remark this answers is not in the document]
Comment (shadowusr) [reply lib/test-attempt-manager.ts:15]: Да, поправил, спасибо!

File: hermione.ts [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+        hermione.on(hermione.events.TEST_FAIL, testResult => {
+            promises.push(queue.add(() => failHandler(testResult).then(addFail)).catch(reject));
+        });
+
+        hermione.on(hermione.events.TEST_PENDING, testResult => {
+            promises.push(queue.add(() => failHandler(testResult as HermioneTestResult).then((testResult) => reportBuilder.addSkipped(testResult)).catch(reject)));
+        });
+
+        hermione.on(hermione.events.RUNNER_END, () => {
+            return Promise.all(promises).then(() => resolve(), reject);
```
[the remark this answers is not in the document]
Comment (shadowusr) [reply hermione.ts:88]: Поправил и словил ошибку — вспомнил, почему так написал =)
Потому что Promise.all возвращает массив результатов, а мы тут ожидаем Promise<void>. Поэтому приходится явно зарезолвить с void.

File: lib/static/modules/reducers/tree/index.js [hunk 10]
```diff
@@ -37,6 +37,7 @@ export default ((state, action) => {
 
             updateAllSuitesStatus(tree, filteredBrowsers);
             initNodesStates({tree, view: state.view});
+            // TODO: handle "updated" status here
```
Comment (shadowusr) [inline lib/static/modules/reducers/tree/index.js:40]: Тут в целом какая концепция:
- серверная сторона записывает в базу статус `updated`
- на клиенте мы получаем этот статус и пытаемся преобразовать в настоящий (например, если тест содержит ошибку, прокрашиваем его в `error`

То есть на сервере у нас везде `updated` статус, а на клиенте — более точный.

Причины:
- Не так просто на сервере узнать настоящий статус теста при accept/undo
- Сервер поставляет больше данных == в любой момент можно поменять отображение

File: lib/plugin-adapter.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (sipayRT) [reply lib/plugin-adapter.ts]: > использоваться в других местах

он сейчас используется в других местах в отрыве от `reportBuilder`? (кроме тестов)

> гибкость замены на другую реализацию

но мы ведь не теряем здесь в гибкости замены, т.к. под капотом ты можешь использовать любой модуль (как и снаружи)

File: hermione.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (shadowusr) [reply hermione.ts:116]: > У нас есть конкретная функциональность reportBuilder-а - регистрация попыток

Вот у нас на это с тобой похоже разное видение. Я так вижу, что зона ответственности reportBuilder — обработка результатов теста (`ReporterTestResult`). Зона ответственности TestAttemptManager — быть реестром попыток и отслеживать количество ретраев.

И эти зоны между собой не пересекаются, потому что attemptManager нужен на этапе создания результата, а reportBuilder — на этапе обработки полностью подготовленного объекта.

Поэтому если мы не можем полностью скрыть от пользователя attemptManager, значит он должен быть полностью вне зоны reportBuilder.

И поразмышляв сейчас я бы его вообще убрал из reportBuilder. Либо думать в сторону полного скрытия attemptManager, но нужно решить проблему, как создать ReporterTestResult не зная попытки.

**Review by sipayRT — APPROVED** [review]
✅ refactor: get rid of plugin-adapter

File: lib/test-adapter/utils/index.ts [hunk 11]
```diff
@@ -1,5 +1,42 @@
...
+        'timestamp',
+        'url'
+    ] as const;
+
+    // Type-level check that we didn't forget to include any keys
+    type A = TupleToUnion<typeof keys>;
+    type B = keyof ReporterTestResult;
+
+    const keysTypeChecked: B extends A ?
+        B extends A ? typeof keys : never
```
Comment (sipayRT) [inline lib/test-adapter/utils/index.ts:38]: тут B и A местами перепутаны

**Review by sipayRT — APPROVED** [review]

File: lib/plugin-adapter.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (shadowusr) [reply lib/plugin-adapter.ts]: Занес создание testAttemptManager в reportBuilder и в целом скрыл его от внешнего использования.

File: hermione.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (shadowusr) [reply hermione.ts:116]: В итоге остановился на таком варианте:

```

const formattedResultWithoutAttempt = formatTestResult(testResult, status, UNKNOWN_ATTEMPT, reportBuilder);

const formattedResult = reportBuilder.addFail(formattedResultWithoutAttempt);
```

То есть testAttemptManager полностью инкапсулирован в reportBuilder и никакие методы не торчат наружу.
